### PR TITLE
store/tikv: parallel commit.

### PR DESCRIPTION
**Pull request by disksing**
@shenli @ngaut @coocood

### CODE REVIEW

**Comment by coocood** [comment]
After the primary region has committed, can we just return and let other region do commit work in background goroutine?

**Comment by shenli** [comment]
LGTM

**Comment by disksing** [comment]
@coocood PTAL
@shenli The change is kind of big. Please help reviewing again :)

File: store/tikv/txn_committer.go [hunk 1]
```diff
@@ -84,51 +84,70 @@ func (c *txnCommitter) primary() []byte {
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	firstIsPrimary := bytes.Equal(keys[0], c.primary())
 
+	var batches []batchKeys
 	// Make sure the group that contains primary key goes first.
-	if primaryRegionID.id != 0 {
-		if err := f(primaryRegionID, groups[primaryRegionID]); err != nil {
+	if firstIsPrimary {
```
Comment (coocood) [inline store/tikv/txn_committer.go]: When will `firstIsPrimary` be false?

File: store/tikv/txn_committer.go [hunk 2]
```diff
@@ -84,51 +84,70 @@ func (c *txnCommitter) primary() []byte {
...
-			err = errors.Trace(e)
+	if len(batches) == 0 {
+		return nil
+	} else if len(batches) == 1 {
+		err = errors.Trace(f(batches[0]))
+		if err != nil {
+			log.Warnf("txnCommitter doBatches failed: %v, tid: %d", err, c.startTS)
+		}
+	} else {
+		// TODO: Stop sending other requests after receiving first error.
```
Comment (coocood) [inline store/tikv/txn_committer.go]: For non-primary regions, we don't have to stop sending other requests after receiving first error.

File: store/tikv/txn_committer.go (hunk 1 above)
Comment (disksing) [inline store/tikv/txn_committer.go]: It happens when we got a RegionError while sending a batch (which may not contain the primary key), then we have to call `iterKeys` to re-split the batch.

File: store/tikv/txn_committer.go (hunk 2 above)
Comment (coocood) [inline store/tikv/txn_committer.go]: Well, if the operation is prewrite, we should stop on first error, but for commit and cleanup, we should not stop.

Comment (disksing) [inline store/tikv/txn_committer.go]: @coocood We may do it for `prewrite`. In prewrite phase, conflict in any single region will abort the whole txn, not matter it contains primary key or not.

Comment (disksing) [inline store/tikv/txn_committer.go]: @coocood Right. I'll fix the comment.

**Comment by coocood** [comment]
LGTM

File: store/tikv/txn_committer.go [hunk 3]
```diff
@@ -297,20 +329,25 @@ func (c *txnCommitter) Commit() error {
...
-			if err := f(id, keys[start:end]); err != nil {
-				return errors.Trace(err)
-			}
+// batchKeys is a batch of keys in the same region.
+type batchKeys struct {
+	region RegionVerID
+	keys   [][]byte
+}
+
+// appendBatchBySize append keys to []batchKeys. It may split the keys to make
```
Comment (shenli) [inline store/tikv/txn_committer.go]: appends

**Comment by shenli** [comment]
Rest LGTM

File: store/tikv/txn_committer.go [hunk 4]
```diff
@@ -82,42 +84,70 @@ func (c *txnCommitter) primary() []byte {
...
+		// TODO: For prewrite, stop sending other requests after receiving first error.
+		ch := make(chan error)
+		for _, batch := range batches {
+			go func(batch batchKeys) {
+				ch <- f(batch)
+			}(batch)
+		}
+		for i := 0; i < len(batches); i++ {
+			if e := <-ch; e != nil {
+				log.Warnf("txnCommitter doBatches failed: %v, tid: %d", err, c.startTS)
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: err -> e ?

**Comment by disksing** [comment]
PTAL @zimulala

**Comment by zimulala** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**